spec/system/admin/reports/sales_tax/sales_tax_totals_by_order_spec.rb

```diff
@@ -425,56 +428,77 @@
-        end
-
-        it 'downloads the file' do
-          expect(downloaded_filenames.length).to eq(0) # downloads folder should be empty
-          select "Spreadsheet", from: "report_format"
-          find("#display_summary_row").uncheck
-          click_on "Go"
-          wait_for_download
-          expect(downloaded_filenames.length).to eq(1) # downloads folder should contain 1 file
-          expect(downloaded_filename).to match(/.*\.xlsx/)
-          downloaded_content = extract_xlsx_rows(downloaded_filename, 1..5)
-          fixture_content = extract_xlsx_rows(report_file_xlsx, 1..5)
-          expect(downloaded_content).to eq(fixture_content)
-        end
-
-        def extract_xlsx_rows(file, range)
-          xlsx = Roo::Excelx.new(file)
-          range.map { |i| xlsx.row(i) }
+      shared_examples "reports generated as" do |output_type, extension|
+        context output_type.to_s do
+          it "downloads the file" do
+            select output_type, from: "report_format"
+
+            expect {
+              generate_report
+            }.to change {
+                   downloaded_filenames.length
+                 }.from(0).to(1)
+            expect(downloaded_filename).to match(/.*\.#{extension}/)
+
+            if extension.eql?("csv")
+              downloaded_file_txt = CSV.read(downloaded_filename).join(" ")
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).not_to match /TOTAL\s+215.0\s+8.61\s+223.61/
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).not_to match /TOTAL\s+115.0\s+4.61\s+119.61/
+            end
+
+            if extension.eql?("xlsx")
+              def extract_xlsx_rows(file, range)
+                xlsx = Roo::Excelx.new(file)
+                range.map { |i| xlsx.row(i) }
+              end
+              downloaded_file_txt = extract_xlsx_rows(downloaded_filename, 1..7).join(" ")
+
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).to match /TOTAL\s+215.0\s+8.61\s+223.61/
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).to match /TOTAL\s+115.0\s+4.61\s+119.61/
+            end
+
+            if extension.eql?("pdf")
+              downloaded_file_pdf = PDF::Reader.new(downloaded_filename)
+
+              # Load PDF pages and contents join into one big string
+              downloaded_file_txt = downloaded_file_pdf.pages.map(&:text).join(" ")
+
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).to match /TOTAL\s+215.0\s+8.61\s+223.61/
+              # correct totals with white-space between them
+              expect(downloaded_file_txt).to match /TOTAL\s+115.0\s+4.61\s+119.61/
+            end
+
+            puts "**DEBUG** downloaded_file_txt:\n #{downloaded_file_txt}"
+
```

Comment:
I'm sure this helps while developing but I think that it shouldn't be in the final code. You can provide a custom fail message like in this example: https://stackoverflow.com/questions/55830876/rspec-how-to-add-more-info-to-the-failure-error-where-it-occurs

But that would be for every assertion. Alternatively, you could rescue the expectation error, print this debug output and then re-raise the error again to make it fail properly.

Reply:
This is good to know, thanks for that. I've removed that `debug` output now.